File: Chapter 04/4.4 Anti-Patterns/After-RichDomain/After/After.DataAcesss/Repositories.cs
using After.DomainModels;

namespace After.DataAccess
{
    /// <summary>
    /// ARCHITECTURE NOTE: By isolating SQL logic here, we prevent 
    /// database concerns from "leaking" into the Presentation or 
    /// Business layers.
    /// </summary>
    // Concrete implementation for a SQL database (simulated)
   public class SqlOrderRepository : IOrderRepository
    {
        public Order GetById(int orderId) { return null; }
        public void Save(Order order) { /* SQL Logic */ }
    }

    public class SqlCustomerRepository : ICustomerRepository
    {
        public Customer GetById(int customerId)
        {
            return new Customer { 
                Id = customerId, Type = "Gold", Email = "a@b.com" 
            };
        }
    }
}

File: Chapter 04/JS/section_4_4_anti_patterns/after_rich_domain_thin_controller/infrastructure/repositories.js
const { IEmailService } = require('./dataAccessInterfaces');

/**
 * ARCHITECTURE NOTE: By isolating Email logic here, we prevent 
 * database concerns from "leaking" into the Presentation or 
 * Business layers.
 * * @implements {IEmailService}
 */
class SmtpEmailService extends IEmailService {
    // Concrete implementation for an email provider
    send(to, subject, body) {
        // Implementation logic would go here
    }
}

/**
 * DATA ACCESS LAYER: SQL IMPLEMENTATION
 * Simulates a database lookup to ensure we get the official, secure price.
 */
class SqlItemRepository {
    getById(itemId) {
        console.log(`  [DB] Fetching official data for Item ID: ${itemId} from SQL.`);
        
        const item = new Item();
        if (itemId === 1) {
            item.price = 100.0;
        } else if (itemId === 2) {
            item.price = 50.0;
        } else {
            item.price = 75.0; // Fallback
        }
        return item;
    }
}

module.exports = { SmtpEmailService, SqlItemRepository };